README.md
<img src="https://raw.githubusercontent.com/Chhekur/local-browser/master/assets/icons/icon.png" width="15%">

# Local Browser

# What is Local Browser ?
Local Browser is a browser built on Electron.
Local Browser is peer to peer browser, which provide you facility to share you localhost with other local browser user's in just one click.

    

# New Features! 🌟

  - ✅ Peer To Peer Browsing
  - ✅ Tab Management
  - ✅ Easy UI

# What Local Browser Does ? 🌟
Local Browser works on **ptp** protocol, ptp stands for **Peer To Peer Protocol**. With Local Browser you don't need to save you website or web application on server inorder to share it, you can keep your files on your machine and can share them with others. Your web app only shared with other users till local browser has opened on your machine. 

### Libraries 🌟

Local Browser uses a number of API's and libraries to work properly:

* ✅ [NodeJs] - evented I/O for the backend
* ✅ [localtunnel] - library to tunnel localhost

### Installation 🌟
> Installers are available on website [here](https://chhekur.github.io/local-browser/).
> Electron requires [Node.js](https://nodejs.org/) to run.



```sh
$ cd local-browser
$ npm install
$ electron . or npm start
```
License
----

MIT

[//]: # "These are reference links used in the body of this note and get stripped out when the markdown processor does its job. There is no need to format nicely because it shouldn't be seen. Thanks SO - http://stackoverflow.com/questions/4823468/store-comments-in-markdown-syntax"


CHANGELOG.md
## Changelog
- Fixed: Static files can be opened
- Fixed: jQuery is not properly loading for some websites

main.js
const electron = require('electron')
global.autoUpdater = require("electron-updater").autoUpdater;
const dialog = require('electron').dialog;
require('./assets/js/main_functions.js');


global.app = electron.app

const BrowserWindow = electron.BrowserWindow
const path = require('path')
const url = require('url')

global.mainWindow
function createWindow () {
	// Create the browser window.
	mainWindow = new BrowserWindow({width: 800, height: 600,minHeight:480, minWidth:360 ,show:false, frame:false})
	// and load the index.html of the app.
	mainWindow.loadURL(url.format({
		pathname: path.join(__dirname, '/views/index.html'),
		protocol: 'file:',
		slashes: true
	}))
	// Open the DevTools.
	mainWindow.webContents.openDevTools()
	// Emitted when the window is closed.
	mainWindow.on('closed', function () {
		// Dereference the window object, usually you would store windows
		// in an array if your app supports multi windows, this is the time
		// when you should delete the corresponding element.
		mainWindow = null
	})
	mainWindow.setMenu(null);
  // require('./menu/menu.js');
}

app.on('ready', function(){
	createWindow();
	mainWindow.once('ready-to-show', () => {
      // splash.destroy();
      mainWindow.show();
    });
    // autoUpdater.checkForUpdates();
})

app.on('ready', function()  {
  autoUpdater.checkForUpdatesAndNotify();
});

app.on('window-all-closed', function () {
	// On OS X it is common for applications and their menu bar
	// to stay active until the user quits explicitly with Cmd + Q
	if (process.platform !== 'darwin') {
		app.quit()
	}
})
app.on('activate', function () {
	// On OS X it's common to re-create a window in the app when the
	// dock icon is clicked and there are no other windows open.
	if (mainWindow === null) {
		createWindow()
	}
})


// auto updates


autoUpdater.on('error', (error) => {
  console.log(error);
  // mainWindow.webContents.send('updateError', error.toString());
  // dialog.showErrorBox('Error: ', error == null ? "unknown" : (error.stack || error).toString())
});

autoUpdater.on('update-available', (info) => {
  // console.log('update available called');
  // console.log(info);
  // mainWindow.webContents.send('updateAvailable',info.version);
  dialog.showMessageBox({
    type: 'info',
    title: 'Found Updates',
    message: 'Found updates, do you want update now?',
    buttons: ['Sure', 'No']
  }, (buttonIndex) => {
    if (buttonIndex === 0) {
        autoUpdater.downloadUpdate();
      //   progressBar = new ProgressBar({
      //   indeterminate: false,
      //   text: 'Preparing data...',
      //   detail: 'Wait...'
      // });
    }
  })
})

// autoUpdater.on('update-not-available', () => {
  // if(isUpdatCallFromMenu){
  //   dialog.showMessageBox({
  //     title: 'No Updates',
  //     message: 'Current version is up-to-date.'
  //   })
  // }
  // mainWindow.webContents.send('updateNotAvailable');
  // updater.enabled = true
  // updater = null
// })

autoUpdater.on('update-downloaded', () => {
  // mainWindow.webContents.send('updateDownloaded');
  dialog.showMessageBox({
    title: 'Install Updates',
    message: 'Updates downloaded, application will be quit for update...'
  }, () => {
    setImmediate(() => autoUpdater.quitAndInstall())
  })
});

autoUpdater.on('download-progress', (progressObj) => {
  // console.log('downloading...');
  // if(mainWindow) mainWindow.webContents.send('downloadProgress', progressObj.percent);
    // progr = progressObj.percent;
    // progress(progressBar);
  let log_message = "Download speed: " + progressObj.bytesPerSecond;
  log_message = log_message + ' - Downloaded ' + progressObj.percent + '%';
  log_message = log_message + ' (' + progressObj.transferred + "/" + progressObj.total + ')';
  console.log(log_message);
})
// autoUpdater.on('update-downloaded', (info) => {
//   console.log('Update downloaded');
// });



assets/js/error_page.js
var errorPageCSS = `
* {
  box-sizing: border-box;
}
a {
  text-decoration: none;
  color: inherit;
  cursor: pointer;
}
body {
  background: #fff;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Ubuntu, Cantarell, "Oxygen Sans", "Helvetica Neue", sans-serif;
}
.btn {
  display: inline-block;
  cursor: pointer;
  color: #777;
  border-radius: 2px;
  background: #fafafa;
  border: 1px solid #ddd;
  font-size: 12px;
  font-weight: 500;
  height: 25px;
  line-height: 2;
  padding: 0 8px;
  letter-spacing: .2px;
  height: 26px;
  font-weight: 400;
}
.btn * {
  cursor: pointer;
  line-height: 25px;
  vertical-align: baseline;
  display: inline-block;
}
.btn:focus {
  outline-color: #007aff;
}
.btn:hover {
  text-decoration: none;
  background: #f0f0f0;
}
.btn.disabled,
.btn:disabled {
  cursor: default;
  color: #999999;
  border: 1px solid #ccc;
  box-shadow: none;
}
.btn.disabled .spinner,
.btn:disabled .spinner {
  color: #aaa;
}
.btn.primary {
  -webkit-font-smoothing: antialiased;
  font-weight: 800;
  background: #007aff;
  color: #fff;
  border: none;
  transition: background .1s ease;
}
.btn.primary:hover {
  background: #0074f2;
}
a.btn span {
  vertical-align: baseline;
}
.btn.big {
  font-size: 18px;
  height: auto;
  padding: 0px 12px;
}
a.link {
  color: blue;
  text-decoration: underline;
}
.right {
  float: right;
}
.icon-wrapper {
  vertical-align: top;
  width: 70px;
  font-size: 50px;
  display: inline-block;
  color: #555;

  i {
    margin-top: -3px;
  }
}
.error-wrapper {
  display: inline-block;
  width: 80%;
}
div.error-page-content {
  max-width: 650px;
  margin: auto;
  transform: translateX(-20px);
  margin-top: 30vh;
  color: #777;
  font-size: 14px;
}
div.error-page-content .description {

  p {
    margin: 20px 0;
  }
}
div.error-page-content i {
  margin-right: 5px;
}
h1 {
  margin: 0;
  color: #333;
  font-weight: 400;
  font-size: 22px;
}
p.big {
  font-size: 18px;
}
.icon {
  float: right;
}
li {
  margin-bottom: 0.5em;
}
li:last-child {
  margin: 0;
}
.footer {
  font-size: 14px;
  color: #777;
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  margin-top: 30px;
  padding-top: 10px;
  border-top: 1px solid #ddd;
}
`

module.exports = function (e) {
  var title = 'This site can’t be reached'
  var info = ''
  var icon = 'fa-exclamation-circle'
  var button = '<a class="btn right" href="javascript:window.location.reload()">Try again</a>'
  var errorDescription
  var moreHelp = ''

  if (typeof e === 'object') {
    errorDescription = e.errorDescription || ''
    info = e.errorInfo || ''
    // remove trailing slash
    var origin = e.validatedURL.slice(0, e.validatedURL.length)

    // strip protocol
    if (origin.startsWith('https://')) {
      origin = origin.slice(8)
    } else if (origin.startsWith('http://')) {
      origin = origin.slice(7)
    }

    switch (e.errorCode) {
      case -106:
        title = 'No internet connection'
        info = `<p>Your computer is not connected to the internet.</p><p>Try:</p><ul><li>Resetting your Wi-Fi connection<li>Checking your router and modem.</li></ul>`
        break
      case -105:
        icon = 'fa-frown-o'
        info = `<p>Couldn’t resolve the DNS address for <strong>${origin}</strong></p>`
        break
      case 404:
        icon = 'fa-frown-o'
        title = e.title || 'Page Not Found'
        info = `<p>${e.errorInfo}</p>`
        break
      case -501:
        icon = 'fa-frown-o'
        info = `<p>Couldn’t resolve the DNS address for <strong>${origin}</strong></p>`
        errorDescription = "ERR_NAME_NOT_RESOLVED"
        break
    }
  } else {
    errorDescription = e
  }

  return `
    <html>
      <head>
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
      </head>
      <body>
        <style>${errorPageCSS}</style>
        <link rel="stylesheet" href="../assets/css/all.css">
        <div class="error-page-content">
          <div class="icon-wrapper"><i class="fa ${icon}"></i></div>

          <div class="error-wrapper">
            <h1>${title}</h1>
            <div class="description">
              ${info}
            </div>
            <div class="footer">
              ${errorDescription}
              ${button}
            </div>
          </div>
        </div>
      </body>
    </html>`.replace(/\n/g, '')
}


assets/js/routes_management.js
const ipc = require('electron').ipcMain;
const dfs = require('dropbox-fs')({
    apiKey: 'JNGlu45zHTAAAAAAAAAACtJtnrBpBH85NBIT7Ow0oI82RGptj28bATaQSAd5TRbr'
});
const ajax = require('ajax-request');

ipc.on('make-route', function(event, url, domain_name){
	dfs.readFile('/routes.json', (err, result) => {
		console.log('reading json ');
		console.log(result);
		let data = JSON.parse(result.toString());
		console.log(data);
		data['ptp://www.' + domain_name] = {
			redirect_url : url
		}
		data['ptp://' + domain_name] = {
			redirect_url: url
		}
		data[url] = {
			redirect_url: 'ptp://www.' + domain_name
		}
		data = JSON.stringify(data,null , 2);
		dfs.writeFile('/routes.json', data , function(err, result) {
	    	if(err) throw err;
	    	// console.log(result);
	    	// event.returnValue = {error: false, msg: 'Successful'};
	    })
	})
})

ipc.on('find-original-route', function(event, url, parsed_url){
	dfs.readFile('/routes.json', (err, result) => {
		let data = JSON.parse(result.toString());
		let temp_url = parsed_url.protocol + '//' + parsed_url.hostname;
		if(data[temp_url] == undefined){
			mainWindow.webContents.send('get-original-route',{error:true, final_url:null} );
			// event.returnValue = {error:true, final_url:null}
		}else{
			let original_url = data[temp_url].redirect_url;
			console.log(temp_url, original_url);
			let final_url = url.replace(temp_url, original_url);
			console.log(url, parsed_url, final_url);
			mainWindow.webContents.send('get-original-route',{error :false, final_url:final_url} );
			// event.returnValue = {error :false, final_url:final_url};
		}
		// console.log(final_url);
		// webview.loadURL(final_url);
	});
})


ipc.on('find-route', function(event, url, parsed_url){
	dfs.readFile('/routes.json', (err, result) => {
		let data = JSON.parse(result.toString());
		let temp_url = parsed_url.protocol + '//' + parsed_url.hostname;
		if(data[temp_url] == undefined){
			event.returnValue = {error:true, final_url:null}
		}else{
			let original_url = data[temp_url].redirect_url;
			console.log(temp_url, original_url);
			let final_url = url.replace(temp_url, original_url);
			console.log(url, parsed_url, final_url);
			event.returnValue = {error :false, final_url:final_url};
		}
		// console.log(final_url);
		// webview.loadURL(final_url);
	});
})

ipc.on('is-domain-available', function(event, domain_name){
	dfs.readFile('/routes.json', (err, result) => {
		// console.log('reading json ');
		// console.log(result);
		let data = JSON.parse(result.toString());
		// console.log(data);
		if(data['ptp://www.' + domain_name] != undefined){
			event.returnValue = false;
		}else{
			event.returnValue = true;
		}
	})
})


function cleareUnavailableRoutes(){
	dfs.readFile('/routes.json', function(err, result){
		if(err) throw err;
		let domains = JSON.parse(result.toString());
		let temp_domains = domains;
		for (let domain in domains){
			if(domains[domain].redirect_url && domains[domain].redirect_url.startsWith('http')){
				ajax({
					url:domains[domain].redirect_url
					// error:function(error){
					// 	if(error.responseText == '404'){
					// 		if(temp_domains[domains[domain]]);
					// 			delete temp_domains[domains[domain]];
					// 		if(temp_domains[domain]);
					// 			delete temp_domains[domain];
					// 	}
					// }
				}, function(error, response, body){
					// console.log(domains[domain].redirect_url);
					console.log(body);
					if(body == "404"){
						if(temp_domains[domains[domain].redirect_url]);
							delete temp_domains[domains[domain].redirect_url];
						if(temp_domains[domain]);
							delete temp_domains[domain];
					}
					let data = JSON.stringify(temp_domains, null, 2)
					dfs.writeFile('/routes.json', data , function(err, result) {
				    	if(err) throw err;
				    	// console.log(result);
				    	// event.returnValue = {error: false, msg: 'Successful'};
				    })
				})
			}
		}
		// console.log(temp_domains);
		mainWindow.webContents.send('unavailableRoutesCleared');
	});
}

ipc.on('cleareUnavailableRoutes', function(event){
	cleareUnavailableRoutes();
});

assets/js/index.js
const ipc = require('electron').ipcRenderer;
const path = require('path');
const url_module = require('url');
const dfs = require('dropbox-fs')({
    apiKey: 'JNGlu45zHTAAAAAAAAAACtJtnrBpBH85NBIT7Ow0oI82RGptj28bATaQSAd5TRbr'
});
var localtunnel = require('localtunnel');
var app = require('http').createServer();
// error_page = require('./error_page.js');

let totalOpenedTab = 0;

// require('./key_binding.js');

let webview;
let newTabCount = 0
let tabs = {}
global.currentTab;


// function getUrl(){
// 	console.log(webview.getURL());
// }

function opentab(tab){
	// console.log('tab opened');
	currentTab = tabs[$(tab).data('tab_id')];
	if(currentTab)
		webview = document.getElementById(currentTab.webview_id);
}

function newTab(url){
	console.log(url);
	newTabCount++;
	totalOpenedTab++;
	$('<li id = "tab' + newTabCount + '" class = "nav-tab" onclick = "opentab(this)" data-tab_id = "tab' + newTabCount + '"><a href=""  role="tab" data-toggle="tab" data-target = "#tab_content' + newTabCount + '"><div class = "inline"><img class = "hide" id = "tab_loading' + newTabCount + '" src = "../assets/img/loading.gif" width = 15></img><img width=15 id = "tab_favicon'+ newTabCount +'"></img></div>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<div class = "inline tab-title-bar"><h id = "tab_click' + newTabCount + '"  data-tab_id = "tab' + newTabCount + '">New Tab</h></div>&nbsp;&nbsp;&nbsp;&nbsp;<span onclick = "closeAnyTab(this)" class="close black" data-tab_id = "tab' + newTabCount + '"> x</span></a></li>').insertBefore('#new-tab-button');
	$('#browsers').append('<div class="tab-pane" id = "tab_content' + newTabCount + '">\
						<div class = "url-bar-header">\
							<div class = "url-bar-buttons">\
							<button class="button url-bar-button" onclick = "backCurrentTab()"><span class="mif-arrow-left"></span></button>\
							<button class="button url-bar-button" onclick = "forwardCurrentTab()"><span class="mif-arrow-right"></span></button>\
							<button class="button url-bar-button" onclick = "reloadCurrentTab()"><i class="fas fa-redo-alt"></i></button>\
							</div>\
							<div class="url-bar-container">\
								<input contenteditable id = "url_bar' + newTabCount + '" class = "url-bar">\
							</div>\
							<div class = "inline url-bar-menu-button">\
								<button class = "button url-bar-button menu-button" data-container="body" data-toggle="popover" data-placement="bottom" data-html = "true" id = "menu"><span class="mif-menu mif-2x"></span></button>\
							</div>\
						</div>\
						<webview id="webview' + newTabCount + '" src = "' + url + '" class = "webview" webpreferences="javascript=yes"></webview>\
					</div>');

	tabs['tab' + newTabCount] = {
		id: newTabCount,
		favicon_id:'tab_favicon' + newTabCount,
		tab_loading_id: 'tab_loading' + newTabCount,
		tab_id : 'tab' + newTabCount,
		url_bar_id: 'url_bar' + newTabCount,
		tab_click_id : 'tab_click' + newTabCount,
		tab_content_id: 'tab_content' + newTabCount,
		webview_id: 'webview' + newTabCount
	}

	initListeners(tabs['tab' + newTabCount]);
	$('#' + tabs['tab' + newTabCount].tab_click_id).click();
	initPopover();
	decreaseTabSize();
}


function decreaseTabSize(){
	$('.nav-tab').css('width', `calc(95% / ${totalOpenedTab})`);
	let width = $('.nav-tab').css('width');
	// console.log(width);
	$('.tab-title-bar').css('width', `calc(${width} - 80px)`);
}


function increaseTabSize(){
	$('.nav-tab').css('width', `calc(95% / ${totalOpenedTab})`);
	let width = $('.nav-tab').css('width');
	$('.tab-title-bar').css('width', `calc(${width} - 80px)`);
}

newTab('https://google.com');
$('#tab_click1').click();

function findNextTab(deleted_tab_count){
    for(let i = deleted_tab_count + 1; i <= newTabCount; i++){
        let key = 'tab' + i;
        if(tabs.hasOwnProperty(key)) return i;
    }
    for(let i = deleted_tab_count - 1; i > 0; i--){
        let key = 'tab' + i;
        if(tabs.hasOwnProperty(key)) return i;
    }
    return undefined;
}

function closeCurrentTab(tab){
	totalOpenedTab--;
	// console.log(tab);
	let next_tab_count = findNextTab(tab.id);
	if(next_tab_count != undefined){
		$('#' + tab.tab_id).remove();
		$('#' + tab.tab_content_id).remove();
		delete tabs[tab.tab_id];
		currentTab = tabs['tab' + next_tab_count];
		$('#' + currentTab.tab_click_id).click();
	}else{
		ipc.send('close-app');
	}

	increaseTabSize();
}


function closeAnyTab(tab){
	closeCurrentTab(tabs[$(tab).data('tab_id')]);
}


function openURL(url){
	let temp = url_module.parse(url);
	// console.log(temp.protocol);
	if(temp.protocol == null || temp.protocol == "localhost:") url = 'http://' + url;
	if(temp.protocol == 'ptp:'){
		console.log(temp);
		let final_url = ipc.sendSync('find-route', url, temp);
		if(final_url.error){
			let generated_error_page = error_page({errorCode:-501, validatedURL:url});
			webview.executeJavaScript(`document.body.innerHTML = '';document.write('${generated_error_page}')`);
		}else{
			console.log(final_url);
			// $('#' + currentTab.url_bar_id).val(final_url.final_url);
			webview.loadURL(final_url.final_url);
		}
	}else if(temp.protocol != 'http:' && temp.protocol != "localhost:" && temp.protocol != 'https:' && temp.protocol != 'ptp:' && temp.protocol != null && temp.protocol != 'file:'){
		let generated_error_page = error_page({errorCode:-501, validatedURL:url});
			webview.executeJavaScript(`document.body.innerHTML = '';document.write('${generated_error_page}')`);
	}else{
		// $('#' + currentTab.url_bar_id).val(url);
		// console.log(url);
		webview.loadURL(url);
		// webview.loadURL($('#url').val());
		console.log(webview.getURL());
	}
}


// capture enter pressed at url bar


$('body').on('keypress', '.url-bar', function(e){
	// console.log('Hello')
	// return e.which != 13;
	if(e.which == 13){
		// console.log($('#' + currentTab.url_bar_id).val());
		let url = $('#' + currentTab.url_bar_id).val();
		console.log(url);
		openURL(url);
		return false;
	}else{
		// return true;
		console.log("don't know what happened");
	}
});


function minimize_app(){
	ipc.send('minimize-app');
}

function maximize_app(){
	ipc.send('maximize-app');
}
function close_app(){
	ipc.send('close-app');
}

function backCurrentTab(){
	webview.goBack();
}
function forwardCurrentTab(){
	webview.goForward();
}
function reloadCurrentTab(){

	openURL($('#' + currentTab.url_bar_id).val().toLowerCase());
	// webview.reload();
}

function updateUrlBar(tab){
	let temp_webview = document.getElementById(tab.webview_id);
	let url = url_module.parse(temp_webview.getURL(), true);
	let temp_url = temp_webview.getURL();

	// handeling about page, need to fix in future

	if(url.protocol == "file:" && path.basename(temp_url) == "about.html"){
		$('#' + currentTab.url_bar_id).val('local://about');
		return;
	}
	// console.log(url);
	if(url.query.error == undefined){
		// console.log('Hello');
		ipc.send('find-original-route', temp_url, url);
	}
}

ipc.on('get-original-route', function(event, final_url){
	if(final_url.error){
		// console.log(webview.getURL());
		let temp_url = url_module.parse(webview.getURL(), true);
		console.log(temp_url);
		if(temp_url.query){
			if(temp_url.query.error != "true" && !temp_url.query.errorDesc){
				// console.log('asfasfasdf');
				// console.log(temp_url.query.error);
				// $('#' + currentTab.url_bar_id).html(`<span style = "color:green">${temp_url.protocol}</span>${webview.getURL().replace(temp_url.protocol, '')}`);

				$('#' + currentTab.url_bar_id).val(webview.getURL());
				// $('#' + currentTab.url_bar_id).attr('data-prepend',temp_url.protocol);
			}
		}else{
			$('#' + currentTab.url_bar_id).val(webview.getURL());
			// $('#' + currentTab.url_bar_id).attr('data-prepend',temp_url.protocol);
		}
	}else{
		$('#' + currentTab.url_bar_id).val(final_url.final_url);
		// $('#' + currentTab.url_bar_id).attr('data-prepend',temp_url.protocol);
	}
})

function initListeners(tab){
	let temp_webview = document.getElementById(tab.webview_id);
	temp_webview.addEventListener('did-finish-load', function(){
	// temp_webview.openDevTools();
		updateUrlBar(tab);
		// $('#' + tab.url_bar_id).val(temp_webview.getURL());
		$('#' + tab.tab_click_id).text(temp_webview.getTitle());
	});

	temp_webview.addEventListener('did-fail-load', function(error){
		let generated_error_page = error_page(error);
		temp_webview.executeJavaScript(`document.body.innerHTML = '';document.write('${generated_error_page}')`);
		// console.log(error);
		// temp_webview.loadURL(path.join('file://', __dirname, '..', 'views','error.html?errorDesc=' + error + '&error=true'));
	});

	temp_webview.addEventListener('page-favicon-updated', function(favicon){
		$('#' + tab.favicon_id).attr('src',favicon.favicons[0]);
		// console.log(favicon);
	});

	temp_webview.addEventListener('did-start-loading', function(){
		$('#' + tab.favicon_id).addClass('hide');
		// $('#' + tab.tab_loading_id).addClass('show');
		$('#' + tab.tab_loading_id).removeClass('hide');
	});
	temp_webview.addEventListener('did-stop-loading', function(){
		$('#' + tab.tab_loading_id).addClass('hide');
		// $('#' + tab.tab_loading_id).removeClass('show');
		$('#' + tab.favicon_id).removeClass('hide');
		// $('#' + tab.favicon_id).removeClass('');
	})
}

$('body').on('click', '.menu-button', function(e){
	initPopover();
});

function initPopover(){
	$("[data-toggle=popover]").each(function(i, obj) {
			// console.log(i,obj);

		$(this).popover({
		  html: true,
		  content: function() {
		  	// console.log(this);
		    var id = $(this).attr('id')
		    return $('#popover-content-' + id).html();
		  }
		});

	});
}

// clear error messages on click domain field

$('body').on('click', '.domain_name_field' ,function(){
	// console.log('Hello')
	$($('.domain-error')[1]).html('');
})


// make popovers hide, when click outside

$('body').on('click', function (e) {
	// console.log('Hello');
    $('[data-toggle=popover]').each(function () {
        // hide any open popovers when the anywhere else in the body is clicked
        if (!$(this).is(e.target) && $(this).has(e.target).length === 0 && $('.popover').has(e.target).length === 0) {
            $(this).popover('hide');
        }
    });
});

function share_website(){
	// console.log($($('.domain_name_field')[1]).val());
	let domain_name = $($('.domain_name_field')[1]).val().toLowerCase();
	console.log(domain_name);
	if(isDomainValid(domain_name).error){
		$($('.domain-error')[1]).html(isDomainValid(domain_name).msg);
	}else{
		// console.log(ipc.sendSync('-domain-available', domain_name));
		if(ipc.sendSync('is-domain-available', domain_name)){
			let url_for_tunnel = $('#' + currentTab.url_bar_id).val();
			let parsed_url = url_module.parse(url_for_tunnel, true);
			// console.log(parsed_url.port);
			let port = parsed_url.port;
			let tunnel = localtunnel((port == null)? 80 : port, function(err){
				if(err) console.log(err);
				console.log(tunnel.url);
				ipc.send('make-route', tunnel.url, domain_name);
				// console.log(result);
				$($('.domain-error')[1]).html(`Your website has been shared<br><a onclick = "openURL('ptp://${domain_name}')">ptp://${domain_name}</a>`);
				// newTab(`ptp://${domain_name}`);
			})
		}else{
			$($('.domain-error')[1]).html('Domain not available');
			// console.log($('.domain-error'))
		}
	}
}


// open dev tools

function openDevTools(){
	webview.openDevTools();
}

// check domain name is valid or not

function isDomainValid(domain){
	let domain_name = domain.trim();
	if(!domain_name.endsWith('.com')) return {error:true, msg:'Add .com at the end'};
	if(domain_name.startsWith('www.')) return {error:true, msg:'No need to include www.'};
	if(domain_name.length < 7) return {error:true, msg:'Domain name should be more than 3 characters'};
	return {error:false}

}

ipc.send('cleareUnavailableRoutes');

ipc.on('unavailableRoutesCleared', function(event){
	console.log('routes cleared');
})

function openAbout(){
	console.log('open about');
	console.log(`file://${path.join(__dirname, '..', 'views', 'about.html')}`);
	newTab(`file://${path.join(__dirname, '..', 'views', 'about.html')}`);
}